sanic/handlers.py

```diff
@@ -61,14 +62,20 @@ def lookup(self, exception):
 
         :return: Registered function if found ``None`` otherwise
         """
-        handler = self.cached_handlers.get(type(exception), self._missing)
-        if handler is self._missing:
-            for exception_class, handler in self.handlers:
-                if isinstance(exception, exception_class):
-                    self.cached_handlers[type(exception)] = handler
-                    return handler
-            self.cached_handlers[type(exception)] = None
-            handler = None
+        handler = self.cached_handlers.get(type(exception))
+        if handler:
+            return handler
+
+        exception_class = type(exception)
+        for ancestor in type.mro(exception_class):
+            handler = self.cached_handlers.get(ancestor)
+            if handler:
+                self.cached_handlers[exception_class] = handler
+                return handler
+            if ancestor is BaseException:
+                break
+        self.cached_handlers[type(exception)] = None
```

Comment:
If the purpose of setting `None` on line 77 is to skip the next search, that test on line 66 should probably be `type(exception) in self.cached_handlers` so that None can be returned instead of redoing the search of ancestors.

Reply:
Possibly same issue on line 72.

I would also suggest removing the `get` before testing if it exists and returning ` return self.cached_handlers[type(exception)]` instead.

Reply:
Cleaned it up some.